Add 'io.netty.tryAllocateUninitializedArray' system property which al…

normanmaurer opened the pull request:
…lows to allocate byte[] without memset in Java9+

Motivation:

Java9 added a new method to Unsafe which allows to allocate a byte[] without memset it. This can have a massive impact in allocation times when the byte[] is big. This change allows to enable this when using Java9 with the io.netty.tryAllocateUninitializedArray property when running Java9+.
Please note that you will need to open up the jdk.internal.misc package via '--add-opens java.base/jdk.internal.misc=ALL-UNNAMED' as well.

Modifications:

Allow to allocate byte[] without memset on Java9+

Result:

Better performance when allocate big heap buffers and using java9.

normanmaurer commented:
See also https://shipilev.net/jvm-anatomy-park/7-initialization-costs/

netkins commented on buffer/src/main/java/io/netty/buffer/UnpooledUnsafeHeapByteBuf.java:
```diff
@@ -29,6 +29,10 @@
         super(alloc, initialCapacity, maxCapacity);
     }
 
+    byte[] allocateArray(int initialCapacity) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

netkins commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -99,6 +99,7 @@
     private static final AtomicLong DIRECT_MEMORY_COUNTER;
     private static final long DIRECT_MEMORY_LIMIT;
     private static final ThreadLocalRandomProvider RANDOM_PROVIDER;
+    private static final boolean TRY_ALLOCATE_UNINITIAlIZED_ARRAY;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this constant name to match the regular expression '^([a-z][a-zA-Z0-9]*|[A-Z][A-Z0-9_v]*)$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS00115)

normanmaurer replied:
false positive ?

jasontedor replied:
I think it's because of the lowercase `l` in the middle of `UNINITIAlIZED`?

normanmaurer replied:
Doh ... you are right

fenik17 commented on buffer/src/main/java/io/netty/buffer/PoolArena.java:
```diff
@@ -672,19 +672,23 @@ private void destroyPoolChunkLists(PoolChunkList<T>... chunkLists) {
         }
 
+        private static byte[] newByteArray(int size) {
+            return HAS_UNSAFE ? PlatformDependent.allocateUninitializedArray(size) : new byte[size];
```
a HAS_UNSAFE check not needed here. It's also checked in the allocateUninitializedArray().

normanmaurer replied:
+1

jasontedor commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -105,7 +107,7 @@ public Object run() {
             if (unsafe != null) {
                 final Unsafe finalUnsafe = unsafe;
-                final Object maybeException = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+                Object maybeException = AccessController.doPrivileged(new PrivilegedAction<Object>() {
```
Sorry, I'm missing why you dropped `final` here?

normanmaurer replied:
let me re-add it again. Messed up while refactoring.

jasontedor commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +279,40 @@ public Object run() {
         }
 
+        logger.debug("sun.misc.Unsafe.allocateUninitializedArray: {}",
+                ALLOCATE_ARRAY_METHOD != null ? "available" : "unavailable");
```
Log any exception that caused it to be unavailable?

normanmaurer replied:
+1

trustin commented:
How about making this behavior default and adding a system property that
disables it?

normanmaurer commented:
@trustin not sure if we want to change the default behaviour or not. Thought it may make sense to keep the old behaviour just to be a bit more conservative . WDYT ? Also @nmittler @Scottmitch

trustin commented:
@normanmaurer I think it's OK, because we never guaranteed that a buffer allocated by `ByteBufAllocator` is zeroed out. We may need to leave some note about this 'potentially' breaking change in the release note. I doubt it will be a problem though.

normanmaurer commented:
@trustin also the user will need to open up access as explained in the commit message:

```
Please note that you will need to open up the jdk.internal.misc package via '--add-opens java.base/jdk.internal.misc=ALL-UNNAMED' as well.
```

So maybe its fine as you said...

fenik17 commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +283,74 @@ public Object run() {
+                                .loadClass("jdk.internal.misc.Unsafe");
+                        Method method = internalUnsafeClass.getDeclaredMethod("getUnsafe");
+                        return method.invoke(null);
+                    } catch (ClassNotFoundException e) {
```
Maybe catch only an `Exception` instead of all?

normanmaurer replied:
I like to just be explicit about what to catch and what not.

Scottmitch replied:
If the exception is not expected do we want this to prevent the application from starting? It looks like we don't really care what the exception is, we just won't use `jdk.internal.misc.Unsafe`.

normanmaurer replied:
Imho it should never throw an other exception so I thought it may be good not not blindly catch everything

nmittler commented:
@normanmaurer I agree with @trustin here, I think it's fine

> I think it's OK, because we never guaranteed that a buffer allocated by ByteBufAllocator is zeroed out. We may need to leave some note about this 'potentially' breaking change in the release note. I doubt it will be a problem though.

normanmaurer commented:
@nmittler @trustin ok changed the default to be `true`. PTAL again

trustin commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +283,74 @@ public Object run() {
+            }
+            ALLOCATE_ARRAY_METHOD = allocateArrayMethod;
+
+            logger.debug("jdk.internal.misc.Unsafe.allocateUninitializedArray: {}",
```
Array: -> Array(): ?

normanmaurer replied:
+1... let me fix this

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +283,74 @@ public Object run() {
+                        maybeException = e;
+                    }
+                } else {
+                    allocateArrayMethod = null;
```
nit: initialize allocateArrayMethod to `null` above so you don't need this else?

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +283,74 @@ public Object run() {
+                    allocateArrayMethod = null;
+                }
+            } else {
+                allocateArrayMethod = null;
```
nit: same comment

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -316,6 +388,19 @@ static ByteBuffer allocateDirectNoCleaner(int capacity) {
+            return new byte[size];
+        }
+        try {
+            return (byte[]) ALLOCATE_ARRAY_METHOD.invoke(INTERNAL_UNSAFE, byte.class, size);
```
Do you have benchmarks to show the impact of using reflection here? I'm concerned that we're trading the cost of memset for the cost of a reflective call.

Ideally, we would just call `Unsafe. allocateUninitializedArray` but that mean that we would have to require Java 9 to build.

normanmaurer replied:
See https://github.com/netty/netty/pull/6570#issuecomment-290071182

trustin replied:
Do we build with JDK 8 now? Then we could delegate this into a separate class that uses `MethodHandle`?

normanmaurer commented:
Will add a benchmark 

> Am 28.03.2017 um 17:13 schrieb Nathan Mittler <notifications@github.com>:
> 
> @nmittler commented on this pull request.
> 
> In common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
> 
> > +                });
> +
> +                if (maybeException instanceof Method) {
> +                    try {
> +                        allocateArrayMethod = (Method) maybeException;
> +                        byte[] bytes = (byte[]) allocateArrayMethod.invoke(finalInternalUnsafe, byte.class, 8);
> +                        assert bytes.length == 8;
> +                    } catch (IllegalAccessException e) {
> +                        allocateArrayMethod = null;
> +                        maybeException = e;
> +                    } catch (InvocationTargetException e) {
> +                        allocateArrayMethod = null;
> +                        maybeException = e;
> +                    }
> +                } else {
> +                    allocateArrayMethod = null;
> nit: initialize allocateArrayMethod to null above so you don't need this else?
> 
> In common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
> 
> > +                    try {
> +                        allocateArrayMethod = (Method) maybeException;
> +                        byte[] bytes = (byte[]) allocateArrayMethod.invoke(finalInternalUnsafe, byte.class, 8);
> +                        assert bytes.length == 8;
> +                    } catch (IllegalAccessException e) {
> +                        allocateArrayMethod = null;
> +                        maybeException = e;
> +                    } catch (InvocationTargetException e) {
> +                        allocateArrayMethod = null;
> +                        maybeException = e;
> +                    }
> +                } else {
> +                    allocateArrayMethod = null;
> +                }
> +            } else {
> +                allocateArrayMethod = null;
> nit: same comment
> 
> In common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
> 
> > @@ -316,6 +388,19 @@ static ByteBuffer allocateDirectNoCleaner(int capacity) {
>          return newDirectBuffer(UNSAFE.allocateMemory(capacity), capacity);
>      }
>  
> +    static byte[] allocateUninitializedArray(int size) {
> +        if (ALLOCATE_ARRAY_METHOD == null) {
> +            return new byte[size];
> +        }
> +        try {
> +            return (byte[]) ALLOCATE_ARRAY_METHOD.invoke(INTERNAL_UNSAFE, byte.class, size);
> Do you have benchmarks to show the impact of using reflection here? I'm concerned that we're trading the cost of memset for the cost of a reflective call.
> 
> Ideally, we would just call Unsafe. allocateUninitializedArray but that mean that we would have to require Java 9 to build.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented:
So I added a jmh benchmark and as @nmittler suspected there is some overhead involved in doing the reflection thingy. So I think after running the benchmark a few times I would suggest we do the allocation based on the size either one way or the other. I seems to pay of for allocations >= 100.

@Scottmitch @trustin @nmittler WDYT ?

```

Result "io.netty.microbench.internal.UnitializedArrayBenchmark.allocateUninitializedByteArray":
  103,443 ±(99.9%) 45,122 ns/op [Average]
  (min, avg, max) = (91,138, 103,443, 127,497), stdev = 16,091
  CI (99.9%): [58,321, 148,564] (assumes normal distribution)


# Run complete. Total time: 00:07:43

Benchmark                                                 (size)  Mode  Cnt     Score      Error  Units
UnitializedArrayBenchmark.allocateInitializedByteArray         1  avgt    6     7,938 ±    1,381  ns/op
UnitializedArrayBenchmark.allocateInitializedByteArray        10  avgt    6     9,046 ±    1,725  ns/op
UnitializedArrayBenchmark.allocateInitializedByteArray       100  avgt    6    27,600 ±    6,358  ns/op
UnitializedArrayBenchmark.allocateInitializedByteArray      1000  avgt    6   176,902 ±   79,175  ns/op
UnitializedArrayBenchmark.allocateInitializedByteArray     10000  avgt    6  1030,709 ±  686,290  ns/op
UnitializedArrayBenchmark.allocateInitializedByteArray    100000  avgt    6  8115,312 ± 1132,039  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray       1  avgt    6    18,466 ±    2,311  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray      10  avgt    6    19,117 ±    2,626  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray     100  avgt    6    22,222 ±   14,708  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray    1000  avgt    6   102,574 ±  147,389  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray   10000  avgt    6   104,798 ±   71,891  ns/op
UnitializedArrayBenchmark.allocateUninitializedByteArray  100000  avgt    6   103,443 ±   45,122  ns/op

Benchmark result is saved to /Users/norman/Documents/workspace/netty/microbench/target/reports/performance/UnitializedArrayBenchmark.json
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 463.698 sec - in io.netty.microbench.internal.UnitializedArrayBenchmark```

normanmaurer commented:
Also @fenik17 ...

normanmaurer commented:
Once we compile with java9 we can remove the reflection thingy.

fenik17 commented:
great! 
So, if it work only for Java 9, should we explicitly check a version?
`TRY_ALLOCATE_UNINITIALIZED_ARRAY = JAVA_VERSION >= 8 && tryAllocateUninitializedArray && hasUnsafe();`

normanmaurer commented:
@fenik17 done

netkins commented on microbench/src/main/java/io/netty/microbench/internal/UnitializedArrayBenchmark.java:
```diff
@@ -0,0 +1,80 @@
+    @Setup(Level.Iteration)
+    public void tearDown() throws Exception {
+        // Trigger a GC and sleep for 2 seconds to give it some time to finish GC
+        System.gc();
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Don't try to be smarter than the JVM, remove this call to run the garbage collector. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1215)

netkins replied:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Don't try to be smarter than the JVM, remove this call to run the garbage collector. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1215)

Scottmitch replied:
was this impacting the benchmark, and is this not something JMH accounts for?

normanmaurer replied:
I did not test if it has an impact removing it. I can tell tho

nmittler commented:
@normanmaurer I think 100 bytes may be a little low and it may not behave the same way on all systems. Maybe playing it safe (no pun intended :)) and bumping the threshold up to something like 1KiB would be better. WDYT?

normanmaurer commented:
@nmittler works for me... @Scottmitch @trustin WDYT ?

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -99,6 +99,7 @@
     private static final AtomicLong DIRECT_MEMORY_COUNTER;
     private static final long DIRECT_MEMORY_LIMIT;
     private static final ThreadLocalRandomProvider RANDOM_PROVIDER;
+    private static final boolean TRY_ALLOCATE_UNINITIALIZED_ARRAY;
```
can we instead make this an `int` which indicates the size to start allocating using unsafe? Something like `<0` means don't use unsafe, `<=0` means use unsafe for arrays of this size or greater. Default of 512b or 1024b based upon the benchmarks sgtm.

normanmaurer replied:
Sure...

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -277,8 +283,69 @@ public Object run() {
+            ALLOCATE_ARRAY_METHOD = allocateArrayMethod;
+
+            logger.debug("jdk.internal.misc.Unsafe.allocateUninitializedArray(): {}",
+                    ALLOCATE_ARRAY_METHOD != null ? "available" : "unavailable", maybeException);
```
the way the logging format is we may not have enough `{}` if `maybeException` is not an exception. Should we only include `maybeException` if an exception occurred and have 2 log statements?

Scottmitch commented on common/src/main/java/io/netty/util/internal/PlatformDependent0.java:
```diff
@@ -316,6 +383,19 @@ static ByteBuffer allocateDirectNoCleaner(int capacity) {
     }
 
+    static byte[] allocateUninitializedArray(int size) {
+        if (ALLOCATE_ARRAY_METHOD == null) {
```
typically `PlatformDependent` takes care of calling methods in `PlatformDependent0` only if appropriate, and `PlatformDependent0` fails if it is called in an incorrectly initialized environment. Can we follow that convention here too? That will avoid duplicate logic checks in `PlatformDependent` and here.

trustin commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -157,9 +158,18 @@ public Random current() {
+        boolean tryAllocateUninitializedArray =
+                SystemPropertyUtil.getBoolean("io.netty.tryAllocateUninitializedArray", true);
+        TRY_ALLOCATE_UNINITIALIZED_ARRAY = tryAllocateUninitializedArray && hasUnsafe() && javaVersion() >= 9;
+        logger.debug("io.netty.tryAllocateUninitializedArray: {}", tryAllocateUninitializedArray);
```
Not directly related with this PR, but weren't we used to prepend `-D` to system property names when logging them?

normanmaurer commented:
@trustin yes we build with java 8.... Good idea about the `MethodHandle`! Let me check as I never used this before :)

normanmaurer commented:
@trustin let me just merge it with reflection for now. We may want to consider using method handles for a few things in a followup pr.

normanmaurer commented:
@Scottmitch @trustin @nmittler ok I addressed all comments and used `1024` as default threshold. Also I removed the `System.gc` from the benchmark as it seems to be fine even without it. PTAL again.

fenik17 commented:
> Good idea about the MethodHandle! 

Or even better to use LambdaMetafactory.

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -155,11 +156,22 @@ public Random current() {
+                SystemPropertyUtil.getInt("io.netty.tryAllocateUninitializedArray", 1024);
+        TRY_ALLOCATE_UNINITIALIZED_ARRAY = javaVersion() >= 9 && PlatformDependent0.hasAllocateArrayMethod() ?
+                tryAllocateUninitializedArray : -1;
+        logger.debug("-Dio.netty.tryAllocateUninitializedArray: {}", tryAllocateUninitializedArray);
```
Do you want to log `tryAllocateUninitializedArray` or `TRY_ALLOCATE_UNINITIALIZED_ARRAY` here? Perhaps the latter would be more helpful for debugging? Or just log both?

nmittler replied:
@normanmaurer if you use `Integer.MAX_VALUE` instead of `-1` as the opt-out case (i.e. never have to worry about negative values), then the condition below will become simpler:

```java
return size > TRY_ALLOCATE_UNINITIALIZED_ARRAY ?
    PlatformDependent0.allocateUninitializedArray(size) :
    new byte[size];
```

normanmaurer replied:
yep let me only log TRY_ALLOCATE_UNINITIALIZED_ARRAY

normanmaurer replied:
I think I would prefer to use -1 as this is the value that we usually use to opt-out

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -99,6 +99,7 @@
     private static final AtomicLong DIRECT_MEMORY_COUNTER;
     private static final long DIRECT_MEMORY_LIMIT;
     private static final ThreadLocalRandomProvider RANDOM_PROVIDER;
+    private static final int TRY_ALLOCATE_UNINITIALIZED_ARRAY;
```
nit: maybe rename to make it more clear how it's used? Something like `UNINITIALIZED_ARRAY_ALLOCATION_THRESHOLD`?

normanmaurer replied:
fine by me

Scottmitch replied:
lgtm after this comment is addressed.

nmittler commented on common/src/main/java/io/netty/util/internal/PlatformDependent.java:
```diff
@@ -155,11 +156,22 @@ public Random current() {
+        logger.debug("-Dio.netty.maxDirectMemory: {} bytes", maxDirectMemory);
+
+        int tryAllocateUninitializedArray =
+                SystemPropertyUtil.getInt("io.netty.tryAllocateUninitializedArray", 1024);
```
nit: If you rename as suggested above, you'll probably want to rename the property to be consistent.

normanmaurer commented:
Cherry-picked into 4.1 (e482d933f755d42f2ed0a29776e2599ae07f1806) and 4.0 (ee198f9c35579de4e72240975bb40c05691fca22)
